rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,357 @@
+
+rocALTensorInfo::rocALTensorInfo(
+    std::vector<size_t> dims,
+    RocalMemType mem_type,
+    RocalTensorDataType data_type) : _type(Type::UNKNOWN),
+                                     _dims(dims),
+                                     _mem_type(mem_type),
+                                     _data_type(data_type)
+{
+    _batch_size = dims.at(0);
+    _num_of_dims = dims.size();
+    _data_size = tensor_data_size(data_type);
+    for(unsigned i = 0; i < _num_of_dims; i++)
+        _data_size *= dims.at(i);
+
+    if(_num_of_dims <= 3)
+        _is_image = false;
+}
+
+void rocALTensor::update_tensor_roi(const std::vector<uint32_t> &width, const std::vector<uint32_t> &height)
+{
+    if(_info.is_image())
+    {
+        auto max_dims = _info.max_dims();
+        unsigned max_width = max_dims.at(0);
+        unsigned max_height = max_dims.at(1);
+
+        if (width.size() != height.size())
+            THROW("Batch size of Tensor height and width info does not match")
+
+        if (width.size() != info().batch_size())
+            THROW("The batch size of actual Tensor height and width different from Tensor batch size " + TOSTR(width.size()) + " != " + TOSTR(info().batch_size()))
+
+        for (unsigned i = 0; i < info().batch_size(); i++)
+        {
+            if (width[i] > max_width)
+            {
+                WRN("Given ROI width is larger than buffer width for tensor[" + TOSTR(i) + "] " + TOSTR(width[i]) + " > " + TOSTR(max_width))
+                _info.get_roi()->at(i).x2 = max_width;
+            }
+            else
+            {
+                _info.get_roi()->at(i).x2 = width[i];
+            }
+            if (height[i] > max_height)
+            {
+                WRN("Given ROI height is larger than buffer with for tensor[" + TOSTR(i) + "] " + TOSTR(height[i]) + " > " + TOSTR(max_height))
+                _info.get_roi()->at(i).y2 = max_height;
+            }
+            else
+            {
+                _info.get_roi()->at(i).y2 = height[i];
+            }
+        }
+    }
+}
+
+rocALTensor::~rocALTensor()
+{
+    _mem_handle = nullptr;
+    if(_vx_handle)
+        vxReleaseTensor(&_vx_handle);
+}
+
+//! Converts the Rocal data_type to OpenVX
+vx_enum interpret_tensor_data_type(RocalTensorDataType data_type)
+{
+    switch (data_type)
+    {
+        case RocalTensorDataType::FP32:
+            return VX_TYPE_FLOAT32;
+        case RocalTensorDataType::FP16:
+            return VX_TYPE_FLOAT16;
+        case RocalTensorDataType::UINT8:
+            return VX_TYPE_UINT8;
+        default:
+            THROW("Unsupported Tensor type " + TOSTR(data_type))
+    }
+}
+
+rocALTensor::rocALTensor(const rocALTensorInfo &tensor_info) : _info(tensor_info)
+{
+    _info._type = rocALTensorInfo::Type::UNKNOWN;
+    _mem_handle = nullptr;
+}
+
+int rocALTensor::create_virtual(vx_context context, vx_graph graph)
+{
+    if (_vx_handle)
+        return -1;
+
```

Comment:
there are lots of empty lines in the code. Please remove unnecessary ones following google coding style